errorcodes/errorcodes.go

```diff
@@ -0,0 +1,241 @@
+	if metadata != nil {
+		metadata[ErrorCodesFeatureMetadataKey] = "true"
+	}
+}
+
+func featureEnabled(metadata map[string]string) bool {
+	if _, ok := metadata[ErrorCodesFeatureMetadataKey]; ok {
+		return true
+	}
+	return false
+}
+
+// NewDaprError create a new DaprError using the supplied metadata and ErrorOptions
+// **Note**: As this code is in `Feature Preview`, it will only continue processing
+// if the ErrorCodes is enabled
+func NewDaprError(err error, metadata map[string]string, options ...ErrorOption) *DaprError {
+	// The following condition can be removed once the
+	// Error Codes Feature is GA
+	if !featureEnabled(metadata) {
+		return nil
+	}
+
+	if err == nil {
+		return nil
+	}
+
+	// Use default values
+	de := &DaprError{
+		err:            err,
+		description:    err.Error(),
+		reason:         NoReason,
+		httpCode:       503,
+		grpcStatusCode: codes.Internal,
+	}
+
+	// Now apply any requested options
+	// to overr
+	for _, option := range options {
+		option(de)
+	}
+
+	return de
+}
+
+// Error implements the error interface.
+func (c *DaprError) Error() string {
+	if c.err != nil {
+		return c.err.Error()
+	}
+	return ""
+}
+
+// Unwrap implements the error unwrapping interface.
+func (c *DaprError) Unwrap() error {
+	return c.err
+}
+
+// Description returns the description of the error.
+func (c *DaprError) Description() string {
+	if c.description != "" {
+		return c.description
+	}
+	return c.err.Error()
+}
+
+func (c *DaprError) SetDescription(description string) {
+	c.description = description
+}
+
+func (c *DaprError) SetResourceInfo(resourceInfo *ResourceInfo) {
+	c.resourceInfo = resourceInfo
+}
+
+func (c *DaprError) SetMetadata(md map[string]string) {
+	c.metadata = md
+}
+
+func WithResourceInfo(resourceInfo *ResourceInfo) ErrorOption {
+	f := func(de *DaprError) {
+		de.resourceInfo = resourceInfo
+	}
+	return f
+}
+
+func WithDescription(description string) ErrorOption {
+	f := func(de *DaprError) {
+		de.description = description
+	}
+	return f
+}
+
+func WithMetadata(md map[string]string) ErrorOption {
+	f := func(de *DaprError) {
+		de.metadata = md
+	}
+	return f
+}
+
+// *** GRPC Methods ***
+
+// FromDaprErrorToGRPC transforms the supplied error into
+// a GRPC Status. It assumes if the supplied error
+// is of type DaprError.
+// Otherwise, returns the original error.
+func FromDaprErrorToGRPC(err error) (*status.Status, error) {
+	de := &DaprError{}
+	if errors.As(err, &de) {
+		if st, ese := de.newStatusError(); ese == nil {
+			return st, nil
+		}
+		return nil, err
+	}
+
+	return nil, err
+}
+
+func (c *DaprError) newStatusError() (*status.Status, error) {
+	messages := []protoiface.MessageV1{
+		newErrorInfo(c.reason, c.metadata),
+	}
+
+	if c.resourceInfo != nil {
+		messages = append(messages, newResourceInfo(c.resourceInfo, c.err))
+	}
+
+	ste, stErr := status.New(c.grpcStatusCode, c.description).WithDetails(messages...)
+	if stErr != nil {
+		return nil, stErr
+	}
+
+	return ste, nil
+}
```

Comment:
Can this be pre-computed when we initialize the Error type in `New` so that we don't throw an error down stream when consumers want to get the `*status.Status`?